core/drivers/imx_ele.c

```diff
@@ -0,0 +1,576 @@
+#define ELE_CMD_RNG_OPEN	    0x20
+#define ELE_CMD_RNG_CLOSE	    0x21
+#define ELE_CMD_RNG_GET		    0x22
+#define ELE_CMD_TRNG_STATE	    0xA4
+#define ELE_CMD_GET_INFO	    0xDA
+#define ELE_CMD_DERIVE_KEY	    0xA9
+
+#define IMX_ELE_TRNG_STATUS_READY 0x3
+
+#define ELE_MU_ID  0x2
+#define ELE_MU_IRQ 0x0
+
+#if defined(CFG_MX8ULP)
+#define ELE_MU_DID 0x7
+#define __aligned_ele __aligned(8)
+#elif defined(CFG_MX93)
+#define ELE_MU_DID 0x3
+#define __aligned_ele __aligned(8)
+#else
+#error "Platform not supported"
+#endif
+
+#define SIZE_MSG_32(_msg) size_msg_32(sizeof(_msg))
+
+#define UID_SIZE (4 * sizeof(uint32_t))
+
+register_phys_mem_pgdir(MEM_AREA_IO_SEC, ELE_BASE_ADDR, ELE_BASE_SIZE);
+
+struct get_info_msg_rsp {
+	uint32_t rsp_code;
+	uint16_t soc_id;
+	uint16_t soc_rev;
+	uint16_t lifecycle;
+	uint16_t sssm_state;
+	uint32_t uid[4];
+	uint32_t sha256_rom_patch[8];
+	uint32_t sha256_fw[8];
+} __packed;
+
+struct session_get_device_info_rsp {
+	uint32_t rsp_code;
+	uint32_t user_sab_id;
+	uint32_t chip_uid[4];
+	uint16_t chip_life_cycle;
+	uint16_t chip_monotonic_counter;
+	uint32_t ele_version;
+	uint32_t ele_version_ext;
+	uint8_t fips_mode;
+	uint8_t reserved[3];
+	uint32_t crc;
+} __packed;
+
+struct response_code {
+	uint8_t status;
+	uint8_t rating;
+	uint16_t rating_extension;
+} __packed;
+
+/*
+ * Print ELE response status and rating
+ *
+ * @rsp response code structure
+ */
+static void print_rsp_code(struct response_code rsp __maybe_unused)
+{
+	DMSG("Response status %#"PRIx8", rating %#"PRIx8" (ext %#"PRIx16")",
+	     rsp.status, rsp.rating, rsp.rating_extension);
+}
+
+/*
+ * Print ELE message header
+ *
+ * @hdr message header
+ */
+static void print_msg_header(struct imx_mu_msg_header hdr __maybe_unused)
+{
+	DMSG("Header vers %#"PRIx8", size %"PRId8", tag %#"PRIx8
+	     ", cmd %#"PRIx8,
+	     hdr.version, hdr.size, hdr.tag, hdr.command);
+}
+
+/*
+ * Print full ELE message content
+ *
+ * @msg message
+ */
+static void dump_message(const struct imx_mu_msg *msg __maybe_unused)
+{
+	size_t i = 0;
+	size_t size = msg->header.size;
+	uint32_t *data __maybe_unused = (uint32_t *)msg;
+
+	DMSG("Dump of message %p(%zu)", data, size);
+	for (i = 0; i < size; i++)
+		DMSG("word %zu: %#"PRIx32, i, data[i]);
+}
+
+/*
+ * Return the number of 32 bits words of the given message.
+ *
+ * @cmd command size in byte
+ */
+static size_t size_msg_32(size_t cmd)
+{
+	/* Roundup and add header size */
+	return ROUNDUP_DIV(cmd, sizeof(uint32_t)) + 1;
+}
+
+/*
+ * The CRC for the message is computed xor-ing all the words of the message:
+ * the header and all the words except the word storing the CRC.
+ *
+ * @msg MU message to hash
+ */
+static uint32_t compute_crc(const struct imx_mu_msg *msg)
+{
+	uint32_t crc = 0;
+	uint8_t i = 0;
+	uint32_t *payload = (uint32_t *)msg;
+
+	assert(msg);
+
+	for (i = 0; i < msg->header.size - 1; i++)
+		crc ^= payload[i];
+
+	return crc;
+}
+
+/*
+ * The CRC is the last word of the message
```

Comment:
"Compute message CRC and update CRC in message header"

Reply:
Ok